tapchannel/commitment.go

```diff
@@ -1505,6 +1507,131 @@ func deriveFundingScriptKey(ctx context.Context, addrBook address.Storage,
+	chainParams *address.ChainParams, localShutdownMsg,
+	remoteShutdownMsg cmsg.AuxShutdownMsg, localOutputIndex,
+	remoteOutputIndex uint32,
+	vPktVersion tappsbt.VPacketVersion) ([]*tappsbt.VPacket, error) {
+
+	if len(c.IncomingHtlcAssets.Val.HtlcOutputs) > 0 ||
+		len(c.OutgoingHtlcAssets.Val.HtlcOutputs) > 0 {
+
+		return nil, fmt.Errorf("commitment contains HTLCs, cannot " +
+			"create vPackets")
+	}
+
+	// We group the inputs by asset ID, so we can create a vPacket for each
+	// asset ID. The number of vPackets is dictated by the number of
+	// different asset IDs in the commitment transaction.
+	groupedInputs := tapsend.GroupProofsByAssetID(inputs)
+	vPackets := make(map[asset.ID]*tappsbt.VPacket, len(groupedInputs))
+	for assetID, proofsByID := range groupedInputs {
+		pkt, err := tappsbt.FromProofs(
+			proofsByID, chainParams, vPktVersion,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error creating vPacket: %w",
+				err)
+		}
+
+		vPackets[assetID] = pkt
+	}
+
+	localOutputs := c.LocalOutputs()
+	remoteOutputs := c.RemoteOutputs()
+
+	// We now distribute the outputs to the vPackets.
+	if len(localOutputs) > 0 {
+		for _, output := range localOutputs {
+			pkt, ok := vPackets[output.AssetID.Val]
+			if !ok {
+				return nil, fmt.Errorf("no vPacket found "+
+					"for asset ID %s", output.AssetID.Val)
+			}
+
+			outAsset := output.Proof.Val.Asset
+			vOut, err := assetToInteractiveVOutput(
+				outAsset, asset.V0, localShutdownMsg,
+				localOutputIndex,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("error creating "+
+					"vOutput: %w", err)
+			}
+
+			pkt.Outputs = append(pkt.Outputs, vOut)
+		}
+	}
+
+	if len(remoteOutputs) > 0 {
+		for _, output := range remoteOutputs {
+			pkt, ok := vPackets[output.AssetID.Val]
+			if !ok {
+				return nil, fmt.Errorf("no vPacket found "+
+					"for asset ID %s", output.AssetID.Val)
+			}
+
+			outAsset := output.Proof.Val.Asset
+			vOut, err := assetToInteractiveVOutput(
+				outAsset, asset.V0, remoteShutdownMsg,
+				remoteOutputIndex,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("error creating "+
+					"vOutput: %w", err)
+			}
+
+			pkt.Outputs = append(pkt.Outputs, vOut)
+		}
+	}
+
+	return maps.Values(vPackets), nil
+}
+
+// assetToInteractiveVOutput creates a VOutput for an asset that is part of an
+// interactive transaction.
+func assetToInteractiveVOutput(a asset.Asset, version asset.Version,
+	shutdownMsg cmsg.AuxShutdownMsg,
+	anchorOutputIndex uint32) (*tappsbt.VOutput, error) {
+
+	scriptKey, ok := shutdownMsg.ScriptKeys.Val[a.ID()]
+	if !ok {
+		return nil, fmt.Errorf("no script key for asset %s", a.ID())
+	}
+
+	var proofDeliveryUrl *url.URL
+	err := lfn.MapOptionZ(
+		shutdownMsg.ProofDeliveryAddr.ValOpt(), func(u []byte) error {
+			var err error
+			proofDeliveryUrl, err = url.Parse(string(u))
+			return err
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode proof delivery "+
+			"address: %w", err)
+	}
```

Comment:
you can avoid the side effect `proofDeliveryUrl` write here by returning `Result` from `MapOptionZ`. We have examples of that in our code.

Reply:
Makes sense, done.